[stable9.1] Skip null groups in group manager

PVince81 opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
Skip null groups in group manager, because data inconsistency could cause some backends (LDAP?) to return null groups.

## Related Issue


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@felixboehm @jvillafanez @butonic

Commit:
Skip null groups in group manager

PVince81 commented:
Steps to reproduce:

Problem can be reproduced with memcache/redis by manually causing a cache inconsistency by meddling with the keys:

1. Setup LDAP
1. Setup OC to use Redis as memcache
1. Create a user "zombie300"
1. Add "zombie300" in two groups "Box100" and "Box105" (I used "memberUid" mode)
1. `flushall` in redis to clear all keys, just in case
1. PROPFIND as zombie300 on the root as this user: `curl -X PROPFIND -u zombie300:zombie300 http://localhost/owncloud/remote.php/webdav/` (this is important to not cause web UI polling to repopulate the cache while testing)
1. Compute key name on CLI: `php -r 'print(md5("groupExists" . "Box105"));'` => f083920ef775eb16207405c1c4dc57e9
1. In Redis CLI, find the: `keys *f083920ef775eb16207405c1c4dc57e9*` => "6a3a92343f0dc7ff3b548564acfcbb91/LDAP-user_ldap--f083920ef775eb16207405c1c4dc57e9" (you will get a different result depending on OC instance id)
1. Go to LDAP and delete the group "Box105"
1. In Redis CLI, delete the key from above: `del 6a3a92343f0dc7ff3b548564acfcbb91/LDAP-user_ldap--f083920ef775eb16207405c1c4dc57e9`
1. Redo the PROPFIND

Note: the cache key can also exist if the group does not exist, its contents will be the value "false" in JSON format.

Before the fix:
```
{"reqId":"HU2rhjW37re7hoEcU0vx","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to a member function getGID() on null at \/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/Principal.php#152","level":3,"time":"2016-12-23T10:53:02+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/","user":"5adb9728-5d44-1036-95d9-3f869d76dba2"}
```

After the fix, the warning from this PR is logged:
```
{"reqId":"eSDUQp4TuoeKtPiYBbS7","remoteAddr":"127.0.0.1","app":"core","message":"Warning: user \"5adb9728-5d44-1036-95d9-3f869d76dba2\" belongs to deleted group: \"Box104\"","level":0,"time":"2016-12-23T10:56:03+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/","user":"5adb9728-5d44-1036-95d9-3f869d76dba2"}
```

PVince81 commented:
So I'd say this PR is mostly about hardening against such inconsistencies until we figure out why they are happening in the first place. Here's a starting point for LDAP's crappy caching: https://github.com/owncloud/user_ldap/issues/30

Please review @DeepDiver1975 @jvillafanez @PhilippSchaffrath @felixboehm @butonic

cdamken commented on lib/private/Group/Manager.php:
```diff
@@ -245,7 +245,12 @@ public function getUserIdGroups($uid) {
+					if (!is_null($aGroup)) {
+						$groups[$groupId] = $aGroup;
+					} else {
+						\OC::$server->getLogger()->debug('Warning: user "' . $uid . '" belongs to deleted group: "' . $groupId . '"', array('app' => 'core'));
```
Would not be better instead debug  a higher log level?  I would suggest 2

felixboehm replied:
debug log level makes sense, with this code it is a designed functionality to handle users in deleted groups.

butonic replied:
The change makes oc gracefully handle null groups. So no warning necessary IMO. But starting the debug log line with 'Warning' is weird. Either the admin should do something to fix this or he shouldn't. If he should we need to tell him whit to do to make the warning go away.

PVince81 replied:
This situation should never happen. The purpose of the warning is to make it possible to find out about such scenarios where the LDAP server (or our LDAP impl) returns inconsistent results.

I'm fine removing the "Warning: " piece of the string.

felixboehm reviewed (approved):
👍

felixboehm commented:
Please add test for deleted groups.

cdamken commented:
> Please add test for deleted groups.

@SergioBertolinSG  Please help here

PVince81 commented:
The automated test will need to interact directly with memcache to be able to delete the key in question there... good luck

Commit:
Also skip null groups in group manager's search function

Commit:
Add more group null checks in sharing code

PVince81 commented:
Also skip groups in "search" function that is used by the users page and others: https://github.com/owncloud/core/pull/26871/commits/400baf3103f8d0636a7b045fbe53ebe239c806f1

Additionally I searched for usages of "groupManager->get(" and found that some parts of the sharing code doesn't properly check for group existence, so I hardened that as well in https://github.com/owncloud/core/pull/26871/commits/8e7e5a9ce94d433426141cd174288351a052f4b3

- [ ] TODO: add unit tests for group manager changes
- [ ] TODO: add unit tests for sharing code changes
- [ ] TODO: integration tests for deleted group cases with existing shares

Commit:
Add unit tests for null group safety in group manager

PVince81 commented:
Seems the share manager hasn't been validating properly (not directly related to this issue, but potential future issues):

- [ ] TODO: fix userCreateChecks to always check user existence
- [ ] TODO: fix groupCreateChecks to always check group existence

Commit:
Add unit tests for sharing code null group checks

PVince81 commented:
Just had a quick test with local DB users and shares: when deleting a group in the users admin page, it automatically deletes all matching shares.

But I think in the case of LDAP the group deletion detection happens with a delay, so it makes sense to harden the sharing code to avoid issues when bumping into shares pointing to deleted groups.
I hope we can find a way to automatically test this with LDAP @davitol

PVince81 commented:
Apparently for LDAP without caching, deleting the group in LDAP has a direct effect in OC and any affected shares disappear in the file listing. However they keep the oc_share entries. This makes the hardening above even more important.
Added test cases here https://github.com/owncloud/user_ldap/issues/35#issuecomment-272218822

PVince81 commented:
Ewww, with redis memcaching enabled (and on this branch) if I delete a group in LDAP, it is still cached. That's ok. The shared mount still appears but the OCS Share API doesn't return the share anymore. So possibly it's reading from a different cache ?!

PVince81 commented:
I debugged into it and the OCS Share API uses `getUserGroups` which still returns the deleted group as it was cached. Then later it calls `inGroup()` which itself has its own cache. But that cache was not populated yet, so it grabbed the value directly from LDAP after the deletion, so it returns false.
So this is the cache discrepancy I described in https://github.com/owncloud/user_ldap/issues/30.

The only thing we can do here is to be prepared to receive inconsistent results. This is what this PR is about.

Ok, then let's continue with the unit tests...

Commit:
Added tests for null groups handling in sharing code

Commit:
Ignore moveShare optional repair in mount provider

In some cases, data is inconsistent in the oc_share table due to legacy
data. The mount provider might attempt to make it consistent but if the
target group does not exist any more it cannot work. In such case we
simply ignore the exception as it is not critical. Keeping the
exception would break user accounts as they would be unable to use
their filesystem.

PVince81 commented:
> integration tests for deleted group cases with existing shares

@SergioBertolinSG do we have integration tests for shares related to deleted groups with the database backend ? (LDAP cases will have to be done in the LDAP app as stated above)

PVince81 commented:
@jvillafanez @butonic @PhilippSchaffrath I need a new review, more code was added.

jvillafanez commented on apps/files_sharing/lib/MountProvider.php:
```diff
@@ -168,7 +168,19 @@ private function buildSuperShares(array $allShares, \OCP\IUser $user) {
+
+						// such issue can usually happen when dealing with
+						// null groups which usually appear with group backend
+						// caching inconsistencies
```
I'd rather log something. It will be easier to debug possible errors or to know what is being ignored.

jvillafanez commented on apps/files_sharing/tests/API/Share20OCSTest.php:
```diff
@@ -521,14 +521,19 @@ public function testCanAccessShare() {
 		$share->method('getSharedWith')->willReturn('group2');
+		$this->assertFalse($this->invokePrivate($this->ocs, 'canAccessShare', [$share]));
 
-		$this->groupManager->method('get')->with('group2')->willReturn($group);
```
I think some adjustments are needed here.

PVince81 replied:
indeed, forgot to use "groupnull"

SergioBertolinSG commented:
> @SergioBertolinSG do we have integration tests for shares related to deleted groups with the database backend ? (LDAP cases will have to be done in the LDAP app as stated above)

No, I don't think so. We have one about orphaned shares, this case is be something similar?

jvillafanez commented on tests/lib/Share20/ManagerTest.php:
```diff
@@ -1123,6 +1123,39 @@ public function testUserCreateChecksIdenticalPathSharedViaGroup() {
+			->with($path)
+			->willReturn([$share2]);
+
+		$this->invokePrivate($this->manager, 'userCreateChecks', [$share]);
```
no assertions in this test?

PVince81 replied:
No assertion possible here. This is similar to other tests.
Basically we run the method and no exception is thrown.

jvillafanez replied:
Checking https://github.com/sebastianbergmann/phpunit-documentation/issues/171#issuecomment-67645583

A `$this->assertNull($this->invokePrivate(....))` might be enough in this case. It's a kind of dummy check that fills the requirements.

PVince81 commented:
> No, I don't think so. We have one about orphaned shares, this case is be something similar?

@SergioBertolinSG similar, but not the same. This is about shares that point to deleted groups. Apparently only reproducible with LDAP + caching so far. Would still be good to have tests that confirm that after deleting a group, the matching shares do not appear neither in the PROPFIND nor in the "shared with you" section (OCS Share API query).

Commit:
Adjust null group handing + tests

PVince81 commented:
@jvillafanez adjusted here https://github.com/owncloud/core/pull/26871/commits/c1eea58b40fce9bffecb0ee1012bb3cf20107157

PVince81 commented:
@jvillafanez added `assertNull()` as requested.

jvillafanez commented:
👍

PVince81 commented:
Forward port to master: https://github.com/owncloud/core/pull/26956
